### PR TITLE
CA-392887: `set_tls_config` immediately after enabling clustering

### PR DESCRIPTION
**Pull request by Vincent-lau**
Previously xapi calls `set_tls_config` after the join call, which will restart the remote server of xapi-clusterd. In the meantime, another xapi-clusterd might also be joining, which causes `distribute_state` to be called while the remote server is restarting.

Now move the `set_tls_config` immediately after the cluster daemon is enabled to reduce the chance of race. Moreover, `join_internal` already creates a tls_config and passes it to xapi-clusterd, there is no point resetting the tls config immediately afterwards.

This does not, however, solve the whole problem. For that, we need to make sure that `distribute_state` and `set_tls_config` cannot happen at the same time. More generally, any remote calls cannot happen while `tls_config` is running. Hence we need them to hold the same lock. This will be done in xapi-clusterd.

### CODE REVIEW

File: ocaml/xapi/xapi_cluster_host.ml [hunk 1]
```diff
@@ -239,15 +239,15 @@ let resync_host ~__context ~host =
               (Ref.string_of self) ;
             Xapi_clustering.Daemon.enable ~__context ;
             maybe_switch_cluster_stack_version ~__context ~self ~cluster_stack ;
+            let verify = Stunnel_client.get_verify_by_default () in
+            set_tls_config ~__context ~self ~verify ;
             (* Note that join_internal and enable both use the clustering lock *)
             Client.Client.Cluster_host.enable ~rpc ~session_id ~self
           ) ;
```
Comment (robhoes) [inline ocaml/xapi/xapi_cluster_host.ml:246]: There is an implicit `else` branch here for the "joined and not enabled" case, which now excludes `set_tls_config` as well. This is correct?

File: ocaml/xapi/xapi_cluster_host.ml [hunk 2]
```diff
@@ -239,15 +239,15 @@ let resync_host ~__context ~host =
...
           (* create the watcher here so that the watcher exists after toolstack restart *)
           create_cluster_watcher_on_master ~__context ~host ;
           Xapi_observer.initialise_observer ~__context
-            Xapi_observer_components.Xapi_clusterd ;
-          let verify = Stunnel_client.get_verify_by_default () in
-          set_tls_config ~__context ~self ~verify
+            Xapi_observer_components.Xapi_clusterd
```
Comment (gangj) [inline ocaml/xapi/xapi_cluster_host.ml:250]: May I know more background of the fix, as I see the plan to fix this issue is:
```
1. move set_tls_config somewhere else, maybe closer to when we start up the cluster daemon

2. when restarting the server (i.e. set_tls_config), we should not allow other remote operations to happen, hence a lock would be useful here
```
Is the 2nd plan considered in this fix?

Comment (Vincent-lau) [reply ocaml/xapi/xapi_cluster_host.ml:250]: no, the second fix is much more complicated. This PR is only for the first fix

File: ocaml/xapi/xapi_cluster_host.ml (hunk 1 above)
Comment (Vincent-lau) [reply ocaml/xapi/xapi_cluster_host.ml:246]: I think this is a good point. If the cluster daemon is not enabled, then in theory xapi-clusterd service itself should not be running, hence any requests to xapi-clusterd should fail. That means the observer initialisation should be moved into the enabled path as well. Need to verify that....

Comment (Vincent-lau) [reply ocaml/xapi/xapi_cluster_host.ml]: I believe there is no need to do anything if xapi-clusterd is not enabled, I have resturctured the code for this

**Comment by Vincent-lau** [comment]
requesting a review from @lindig since he originally wrote this code to make sure this refactor is valid

**Review by lindig — APPROVED** [review]
I believe this is valid. When we wrote TLS certification code we did not think carefully enough about the sequence.

**Review by psafont — APPROVED** [review]

**Comment by Vincent-lau** [comment]
xrt tests pending, let's see wait for the tests before merging this

**Comment by lindig** [comment]
@Vincent-lau how are the tests? Can we merge this?

**Comment by Vincent-lau** [comment]
Let's draft this as it might take some time to triage the xrt issues

**Comment by Vincent-lau** [comment]
A few small refactoring to fix bugs in xrt, should be ready to go now.

**Pull request merged**